# Statistical Methods in Pandas - Lab

## Introduction

In this lab you'll get some hands-on experience using some of the key summary statistics methods in Pandas.

## Objectives
You will be able to:

- Calculate summary statistics for a series and DataFrame 
- Use the `.apply()` or `.applymap()` methods to apply a function to a pandas series or DataFrame  


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets. You will find this dataset in the file `'lego_sets.csv'`.   

In the cell below:

- Import Pandas and set the standard alias of `pd`
- Import the `'lego_sets.csv'` dataset 
- Display the first five rows of the DataFrame to get a feel for what we'll be working with

In [41]:
# Import pandas
import pandas as pd


# Import the 'lego_sets.csv' dataset
df = pd.read_csv('lego_sets.csv')
df.head()

# Print the first five rows of DataFrame


ages  list_price  num_reviews  piece_count  play_star_rating  \
0  6-12       29.99          2.0        277.0               4.0   
1  6-12       19.99          2.0        168.0               4.0   
2  6-12       12.99         11.0         74.0               4.3   
3   12+       99.99         23.0       1032.0               3.6   
4   12+       79.99         14.0        744.0               3.2   

                                           prod_desc  prod_id  \
0  Catapult into action and take back the eggs fr...  75823.0   
1  Launch a flying attack and rescue the eggs fro...  75822.0   
2  Chase the piggy with lightning-fast Chuck and ...  75821.0   
3  Explore the architecture of the United States ...  21030.0   
4  Recreate the Solomon R. Guggenheim Museum® wit...  21035.0   

                                      prod_long_desc review_difficulty  \
0  Use the staircase catapult to launch Red into ...           Average   
1  Pilot Pig has taken off from Bird Island with ...              Easy   
2  Pitch speedy bird Chuck against the Piggy Car....              Easy   
3  Discover the architectural secrets of the icon...           Average   
4  Discover the architectural secrets of Frank Ll...       Challenging   

                         set_name  star_rating    theme_name  val_star_rating  \
0           Bird Island Egg Heist          4.5  Angry Birds™              4.0   
1              Piggy Plane Attack          5.0  Angry Birds™              4.0   
2                Piggy Car Escape          4.3  Angry Birds™              4.1   
3  United States Capitol Building          4.6  Architecture              4.3   
4   Solomon R. Guggenheim Museum®          4.6  Architecture              4.1   

  country  
0      US  
1      US  
2      US  
3      US  
4      US

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset. There are two ways we'll get this information -- `.info()` and `.describe()`.

The `.info()` method provides us metadata on the DataFrame itself. This allows us to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset? 
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method. 

In [54]:
# Call the .info() method
df.info()
df['ages'].dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null float64
num_reviews          10641 non-null float64
piece_count          12261 non-null float64
play_star_rating     10486 non-null float64
prod_desc            11884 non-null object
prod_id              12261 non-null float64
prod_long_desc       12261 non-null object
review_difficulty    10206 non-null object
set_name             12261 non-null object
star_rating          10641 non-null float64
theme_name           12258 non-null object
val_star_rating      10466 non-null float64
country              12261 non-null object
dtypes: float64(7), object(7)
memory usage: 1.3+ MB


dtype('O')

#### Interpreting the Results

Read the output above, and then answer the following questions:

- How many total rows are in this DataFrame?  
- How many columns contain numeric data? 
- How many contain categorical data? 
- Identify at least 3 columns that contain missing values. 

Write your answer below this line:
________________________________________________________________________________________________________________________________

- 12261
- 7
- 7
- star_rating, val_star_rating, prod_desc

Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method. 

In [43]:
# Call the .describe() method
df.describe()

list_price   num_reviews   piece_count  play_star_rating  \
count  12261.000000  10641.000000  12261.000000      10486.000000   
mean      65.141998     16.826238    493.405921          4.337641   
std       91.980429     36.368984    825.364580          0.652051   
min        2.272400      1.000000      1.000000          1.000000   
25%       19.990000      2.000000     97.000000          4.000000   
50%       36.587800      6.000000    216.000000          4.500000   
75%       70.192200     13.000000    544.000000          4.800000   
max     1104.870000    367.000000   7541.000000          5.000000   

            prod_id   star_rating  val_star_rating  
count  1.226100e+04  10641.000000     10466.000000  
mean   5.983675e+04      4.514134         4.228960  
std    1.638115e+05      0.518865         0.660282  
min    6.300000e+02      1.800000         1.000000  
25%    2.103400e+04      4.300000         4.000000  
50%    4.206900e+04      4.700000         4.300000  
75%    7.092200e+04      5.000000         4.700000  
max    2.000431e+06      5.000000         5.000000

## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods. 

In the cell below, compute the median value of the `star_rating` column.

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns. Use these to answer the following questions:

- How much is the standard deviation for `piece count`? 
- How many pieces are in the largest lego set?
- How many in the smallest lego set? What is the median `val_star_rating`?

________________________________________________________________________________________________________________________________

- 825.364580	
- 7541
- 1
- 4.3

In [44]:
# Calculate the median of the star_rating column 
# is this not just included in the described data? it is 4.7
df['star_rating'].median()


4.7

Next, get a count of the total number of unique values in `play_star_rating`.

In [45]:
# Print the number of unique values in play_star_rating
df['play_star_rating'].unique()

array([4. , 4.3, 3.6, 3.2, 3.7, 4.4, 4.1, 4.2, 3.8, 4.7, 3. , 5. , 2. ,
       nan, 4.6, 2.7, 4.5, 1. , 3.5, 3.3, 3.9, 4.8, 2.9, 3.4, 4.9, 2.5,
       2.2, 2.3, 2.8, 3.1, 2.1])

Now, compute the standard deviation of the `list_price` column.

In [46]:
# Calculate the standard deviation of the list_price col
import numpy as np

df['prod_id'].unique()

array([7.582300e+04, 7.582200e+04, 7.582100e+04, 2.103000e+04,
       2.103500e+04, 2.103900e+04, 2.102800e+04, 2.102900e+04,
       2.103400e+04, 2.103300e+04, 2.103600e+04, 2.101900e+04,
       2.103200e+04, 1.710100e+04, 4.159700e+04, 4.161400e+04,
       4.161300e+04, 4.160900e+04, 4.160800e+04, 4.160400e+04,
       4.160500e+04, 4.160600e+04, 4.160700e+04, 4.148500e+04,
       4.027000e+04, 4.159900e+04, 4.159800e+04, 4.160300e+04,
       4.160200e+04, 4.160000e+04, 4.160100e+04, 4.161100e+04,
       4.161000e+04, 4.027100e+04, 4.148700e+04, 4.148800e+04,
       4.159400e+04, 4.148600e+04, 4.158500e+04, 4.159000e+04,
       4.159300e+04, 4.159200e+04, 4.159500e+04, 4.159600e+04,
       4.158800e+04, 4.159100e+04, 4.158700e+04, 4.158600e+04,
       6.017500e+04, 6.005200e+04, 6.005100e+04, 6.016200e+04,
       6.016700e+04, 6.016100e+04, 6.014100e+04, 6.018800e+04,
       6.011000e+04, 6.017400e+04, 6.016900e+04, 6.014000e+04,
       6.016000e+04, 6.017300e+04, 6.016600e+04, 6.0139

If we bought every single lego set in this dataset, how many pieces would we have?  

> **Note**: If you truly want to answer this accurately, and are up for the challenge, remove duplicate lego-set entries before summing the pieces. That is, many of the lego sets are listed multiple times in the dataset above, depending on the country where it is being sold and other unique parameters. If you're stuck, just practice calculating the total number of pieces in the dataset for now.

In [55]:
# Total number of pieces across all unique Lego sets
df['piece_count'].sum()
df.select_dtypes(include=['float64', 'int64'])


list_price  num_reviews  piece_count  play_star_rating  prod_id  \
0         29.9900          2.0        277.0               4.0  75823.0   
1         19.9900          2.0        168.0               4.0  75822.0   
2         12.9900         11.0         74.0               4.3  75821.0   
3         99.9900         23.0       1032.0               3.6  21030.0   
4         79.9900         14.0        744.0               3.2  21035.0   
...           ...          ...          ...               ...      ...   
12256     36.5878          6.0        341.0               4.4  70609.0   
12257     24.3878          8.0        217.0               4.1  70629.0   
12258     24.3878         18.0        233.0               4.6  70607.0   
12259     12.1878          1.0         48.0               5.0  70628.0   
12260     12.1878         11.0        109.0               4.5  70606.0   

       star_rating  val_star_rating  
0              4.5              4.0  
1              5.0              4.0  
2              4.3              4.1  
3              4.6              4.3  
4              4.6              4.1  
...            ...              ...  
12256          4.3              4.2  
12257          3.6              4.1  
12258          4.6              4.5  
12259          5.0              5.0  
12260          4.7              4.8  

[12261 rows x 7 columns]

Now, let's try getting the value for the 90% quantile for all numerical columns.  Do this in the cell below.

In [56]:
# Get the 90% quantile for all numerical columns
for col in df.select_dtypes(include=['float64', 'int64']):
    print(np.percentile(df[col], 90))
    

136.2971
nan
1077.0
nan
75531.0
nan
nan


## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data -- there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data. 

In the cell below, print the unique values contained within the `review_difficulty` column. 

In [57]:
# Print the unique values in the review_difficulty column
df['review_difficulty'].unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts()` for this column, to see how common each is. 

In [58]:
# Get the value_counts() of the review_difficulty column
df['review_difficulty'].value_counts()

Easy                4236
Average             3765
Very Easy           1139
Challenging         1058
Very Challenging       8
Name: review_difficulty, dtype: int64

As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `.applymap()` method and passing in a lambda function. 

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call the DataFrame's `.applymap()` method and pass in `lambda x: str(x)`  
* Call the new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [60]:
# Call the .applymap() method
string_df = df.applymap(lambda x: str(x))

# Call the .info() method
string_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null object
num_reviews          12261 non-null object
piece_count          12261 non-null object
play_star_rating     12261 non-null object
prod_desc            12261 non-null object
prod_id              12261 non-null object
prod_long_desc       12261 non-null object
review_difficulty    12261 non-null object
set_name             12261 non-null object
star_rating          12261 non-null object
theme_name           12261 non-null object
val_star_rating      12261 non-null object
country              12261 non-null object
dtypes: object(14)
memory usage: 1.3+ MB


Note that everything -- even the `NaN` values, have been cast to a string in the example above. 

Note that for Pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `.apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame -- this will become especially useful when we learn how to **_normalize_** our datasets in a later section!

## Summary

In this lab, we learned how to:

* Use the `df.describe()` and `df.info()` summary statistics methods 
* Use built-in Pandas methods for calculating summary statistics 
* Apply a function to every element in a DataFrame